### PR TITLE
Introduction of Labeled Property Graph Model & Adjustement of Model Handeling

### PR DESCRIPTION
**Pull request by datomo**
## Summary ( This PR requires the UI PR: [Pull Request](https://github.com/polypheny/Polypheny-UI/pull/75) ) 

This PR introduces a variety of changes into Polypheny. For one it introduces the graph model (labeled property graph model) as new data model, which can be modified and queried like the existing relational and document model. The graph model comes with its own set of algebra nodes and also introduces the `Cypher` query language and `Neo4j` as a native `LPG` data store to fully utilize the `LPG` model. 
Further the PR separates the document model from the relational model by also adding a set of new algebra operations for it.

Additionally the PR introduces additional common algebra operations like the Transformer, which handles the also added `ModelTrait` and the switching between them. 

Finally it fixes different potential bug with existing common operators like the `Streamer` and the `ConstraintEnforcer`.

### Changes

* Renames structures of model independent methods/classes into `Namespace`, `Entity` and `Field` and model specific ones into the model appropriate structural names like Schema, Table and Column for the relational model or `Database`, `Collection` and `Field` for the document model
* Moves document operators into a set of specific operators like `Documents`, `DocumentScan`, etc.
* Enabling and adjusting update logic for document model to use the `Streamer` and `Transformer`, if the model is not supported natively by the underlying store
* Adds new catalog objects and methods to store the graph and document schema
* Adds new `Trait` to the existing available `TraitSet`, which indicates the data model of `AlgNode`, called `ModelTrait`

### Fixes

* Remove some of the book-keeping entities used during optimization of the `AlgNode` as the lead to exponential memory usage
* Gradle dependency logs and unnecessary rebuilding of all modules
* Fixes  https://github.com/polypheny/Polypheny-DB/issues/399
* Fixes https://github.com/polypheny/Polypheny-DB/issues/413

---

### Features

* LPG algebra operators like `LpgScan`, `LpgModify`, etc.
* Cypher query language, which allows to structure, modify and query the LPG data model natively
* Neo4j data store, which for one supports the relational model as well as the newly added LPG model natively
* New `PolySchema` Entities for the LPG model as well as the document model in `LpgGraph` and `DocumentCollection`

* Added crossmodel queries for all 3 available data models, which allows to use a query language, which does not support the data model natively e.g. query a table via `MQL` or querying a `LPG` graph via `SQL`

### Tests

* Set of integration test for Cypher queries, including `DDL`, `DML` and `DQL` queries
* Simple set of integration tests for cross model queries for each combination of the three supported data models

---

### ToDo

- [x] Finalize this Page
- [x] Open UI PR
- [x] Fix DOCUMENT-Editor in UI
- [x] Add logic for each cross model query case

### CODE REVIEW

File: cypher-language/src/test/java/org/polypheny/db/ParserTest.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.polypheny.db;
+
+public class ParserTest {
+
+
```
Comment (vogti) [inline cypher-language/src/test/java/org/polypheny/db/ParserTest.java:21]: Is this class intentionally left empty?

File: cypher-language/build.gradle [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+buildscript {
+    dependencies {
+        classpath group: "gradle.plugin.ca.coglinc2", name: "javacc-gradle-plugin", version: javacc_plugin_version
+    }
+}
+
+
+dependencies {
+    implementation project(":core")
+    implementation project(":sql-language")
```
Comment (vogti) [inline cypher-language/build.gradle:22]: Dependencies between languages might be problematic.

File: catalog/build.gradle [hunk 3]
```diff
@@ -5,6 +5,7 @@ dependencies {
     implementation project(":core")
     implementation project(":sql-language")
     implementation project(":mql-language")
+    implementation project(":jdbc-adapter")
```
[the remark this answers is not in the document]
Comment (datomo) [reply catalog/build.gradle]: Dependency was removed

File: build.gradle [hunk 4]
```diff
@@ -133,8 +134,6 @@ allprojects {
         logLevel 'lifecycle'
     }
 
-    lombok.config['lombok.val.flagUsage'] = 'error'
-    lombok.config['lombok.var.flagUsage'] = 'error'
```
[the remark this answers is not in the document]
Comment (datomo) [reply build.gradle:137]: Lombok does not longer encourage to let gradle create the lombok.config files. They are already present in the master atm. 
I removed them from the submodules and adjusted the one in the root to allow bubbling. This allows to still overwrite the config for submodules if needed but removes the unnecessary files from the modules. This was tested with the val flagUsage, which flags the usage correctly.

File: cypher-language/src/test/java/org/polypheny/db/ParserTest.java (hunk 1 above)
Comment (datomo) [reply cypher-language/src/test/java/org/polypheny/db/ParserTest.java]: Removed

File: cypher-language/build.gradle (hunk 2 above)
Comment (datomo) [reply cypher-language/build.gradle]: Removed dependency, as it was not longer needed.

File: core/src/main/java/org/polypheny/db/algebra/core/document/DocumentValues.java [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+
+    public LogicalValues getRelationalEquivalent() {
+        AlgTraitSet out = traitSet.replace( ModelTrait.RELATIONAL );
+        AlgOptCluster cluster = AlgOptCluster.create( getCluster().getPlanner(), getCluster().getRexBuilder() );
+
+        AlgRecordType rowType = new AlgRecordType( List.of(
+                new AlgDataTypeFieldImpl( "_id_", 0, cluster.getTypeFactory().createPolyType( PolyType.VARCHAR, 24 ) ),
+                new AlgDataTypeFieldImpl( "_data_", 1, cluster.getTypeFactory().createPolyType( PolyType.VARCHAR, 2024 ) ) ) );
```
[the remark this answers is not in the document]
Comment (datomo) [reply core/src/main/java/org/polypheny/db/algebra/core/document/DocumentValues.java:134]: Due to the format of the RowType this is sadly still the way which these entries have to be configured. After the RowType adjustement for the 0.10 release, we should be able to adjust this to a less strict form, which is more in line with the document data model.

**Review by vogti — APPROVED** [review]
Thx, @datomo, for this contribution!

**Pull request merged**